### PR TITLE
Add `iterator.Collect` function

### PR DESCRIPTION
**Pull request by nathan-artie**
Makes testing iterators a lot easier.

### CODE REVIEW

File: lib/iterator/types.go [hunk 1]
```diff
@@ -1,6 +0,0 @@
-package iterator
```
Comment (nathan-artie) [inline lib/iterator/types.go:1]: Renamed this file to `iterator.go`

File: lib/iterator/iterator.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package iterator
+
+type Iterator[T any] interface {
```
Comment (nathan-artie) [inline lib/iterator/iterator.go:3]: This used to be in `types.go`.

File: lib/iterator/batch.go [hunk 3]
```diff
@@ -8,8 +8,8 @@ type batchIterator[T any] struct {
 	step  int
 }
 
-// Returns an iterater that splits a list of items into batches of the given step size.
-func NewBatchIterator[T any](items []T, step int) Iterator[[]T] {
+// Batched returns an iterater that splits a list of items into batches of the given step size.
```
Comment (Tang8330) [inline lib/iterator/batch.go:11]: `iterator`

File: lib/iterator/iterator.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	HasNext() bool
+	Next() (T, error)
+}
+
+// Collect returns a new slice containing all the items from an [Iterator].
+func Collect[T any](iter Iterator[T]) ([]T, error) {
```
Comment (Tang8330) [inline lib/iterator/iterator.go:9]: Is this only for tests?

Comment (nathan-artie) [reply lib/iterator/iterator.go:9]: For now, I can see it being used in the future though.

Comment (Tang8330) [reply lib/iterator/iterator.go:9]: If so, can we make that more clear?

Comment (Tang8330) [reply lib/iterator/iterator.go:9]: Running this in Prod over a large data set can cause OOM

Comment (Tang8330) [reply lib/iterator/iterator.go:9]: I do like this function for testing as it can reduce boilerplate

**Review by Tang8330 — APPROVED** [review]

**Pull request merged**